File: web/src/lib/nano-rpc.ts
import { getEnv } from "./env";

// BlackBird talks to exactly ONE hard-coded Nano RPC endpoint so environment
// variables cannot redirect calls to a different node. There is no fallback:
// rpc.nano-gpt.com was removed 2026-09-11 (expired TLS certificate — it answered
// nothing while masking the real error from nano.to). If rpc.nano.to is down the
// app shows a maintenance screen rather than silently rerouting user traffic.
// Local/backend proof-of-work remains the work fallback.
const NANO_RPC_ENDPOINT = "https://rpc.nano.to";

const DEFAULT_TIMEOUT_MS = 30_000;

export type NanoRpcResponse<T = unknown> =
  | { error: string }
  | T;

/** A real answer from a responsive endpoint (e.g. "Account not found") —
 * never a reason to fail over. */
class SemanticError extends Error {}

/** rpc.nano.to rejected our credential. It answers with HTTP 200 and
 * {"error":"Invalid API Key."}, which otherwise reads as a SemanticError and
 * fails the call outright. The node itself is healthy and its keyless tier
 * serves every read we make, so the right move is to retry the SAME endpoint
 * without the key. */
class AuthError extends Error {}

function isAuthErrorMessage(msg: string): boolean {
  const m = msg.toLowerCase();
  return (
    m.includes("invalid api key") ||
    m.includes("invalid key") ||
    m.includes("unauthorized")
  );
}

/** Latched once nano.to rejects the key so later calls skip the doomed
 * round-trip. Read by /api/greenlight to surface a degraded (not broken) tier. */
let keyRejected = false;

export function isRpcKeyRejected(): boolean {
  return keyRejected;
}

async function callEndpoint<T>(
  endpoint: string,
  action: string,
  params: Record<string, unknown>,
  timeoutMs: number,
  withKey = true
): Promise<T> {
  const env = getEnv();
  // The API key belongs to rpc.nano.to ONLY — never sent to the fallback, and
  // never again once nano.to has rejected it.
  const useKey =
    endpoint === NANO_RPC_ENDPOINT && withKey && !keyRejected && !!env.NANO_RPC_KEY;
  const body = useKey
    ? { action, ...params, key: env.NANO_RPC_KEY }
    : { action, ...params };

  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);
  try {
    const response = await fetch(endpoint, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        ...(useKey ? { Authorization: env.NANO_RPC_KEY } : {}),
        "User-Agent": "BlackBird-web/1.0",
      },
      body: JSON.stringify(body),
      cache: "no-store",
      signal: controller.signal,
    });

    if (!response.ok) {
      throw new Error(`HTTP ${response.status} from ${endpoint}`);
    }

    const data = (await response.json()) as NanoRpcResponse<T>;

    if (data && typeof data === "object" && "error" in data) {
      const message = String(data.error);
      if (useKey && isAuthErrorMessage(message)) {
        throw new AuthError(message);
      }
      throw new SemanticError(`Nano RPC error: ${message}`);
    }

    return data as T;
  } finally {
    clearTimeout(timer);
  }
}

export async function nanoRpcCall<T = unknown>(
  action: string,
  params: Record<string, unknown> = {},
  timeoutMs: number = DEFAULT_TIMEOUT_MS
): Promise<T> {
  try {
    return await callEndpoint<T>(NANO_RPC_ENDPOINT, action, params, timeoutMs);
  } catch (err) {
    if (err instanceof AuthError) {
      // Same endpoint, keyless. Latch so the next call goes straight there.
      keyRejected = true;
      console.warn(
        `nano-rpc: rpc.nano.to rejected NANO_RPC_KEY (${err.message}); ` +
          "continuing keyless. Renew the key to restore the paid tier."
      );
      return await callEndpoint<T>(
        NANO_RPC_ENDPOINT,
        action,
        params,
        timeoutMs,
        false
      );
    }
    if (err instanceof SemanticError) throw new Error(err.message);
    // No fallback node by design: surface the failure so the caller (and the
    // maintenance screen) can react to a real rpc.nano.to outage.
    throw err;
  }
}

export async function getAccountBalance(account: string) {
  const data = await nanoRpcCall<{
    balance?: string;
    pending?: string;
  }>("account_balance", { account });
  return {
    balance: data.balance ?? "0",
    pending: data.pending ?? "0",
  };
}

export async function getAccountInfo(account: string) {
  return nanoRpcCall<{
    frontier?: string;
    open_block?: string;
    representative?: string;
    balance?: string;
    modified_timestamp?: string;
    block_count?: string;
  }>("account_info", {
    account,
    representative: "true",
  });
}

export async function getPendingBlocks(account: string) {
  return nanoRpcCall<{
    blocks?: Record<string, { amount: string; source: string }>;
  }>("pending", {
    account,
    count: 20,
    source: true,
    include_active: true,
  });
}

File: src/nano_rpc.py
"""
Shared Nano RPC client.

Learns from the user's other projects (AiSwarmResearch, fishing, XNO_GAME_Template,
ordinal, xchat-alpha):
- Default endpoint: https://rpc.nano.to
- API key env var: NANO_RPC_KEY (falls back to NANO_RPC_API_KEY)
- Auth header: Authorization: <key> (no Bearer prefix)
- Public fallbacks when a public RPC is configured
- Keep-alive sessions, timeout, retries
- Distinguish transport errors from genuine Nano errors
- Remember last good endpoint
"""
import os
from typing import List, Optional

import requests
from requests.adapters import HTTPAdapter


# rpc.nano.to is the ONE and ONLY endpoint (owner policy, AGENTS.md). There is
# no fallback node: rpc.nano-gpt.com was removed 2026-09-11 (its TLS certificate
# had expired, so it answered nothing while still masking the real error from
# nano.to). When rpc.nano.to is unreachable the correct behaviour is to say so
# loudly — the web app shows a maintenance screen — not to silently reroute
# traffic, and never to send the API key to a third-party node.
PRIMARY_RPC_URL = "https://rpc.nano.to"
DEFAULT_RPC_URLS = [PRIMARY_RPC_URL]

# Nano errors that are valid ledger answers, not endpoint failures.
_NANO_ERRORS = (
    "account not found",
    "block not found",
    "bad account",
    "old block",
    "fork",
    "gap",
    "unreceivable",
    "insufficient balance",
    "unopened",
)

# rpc.nano.to answers a rejected key with HTTP 200 and {"error": "Invalid API
# Key."}. That is neither a ledger answer nor a transport fault: the endpoint is
# healthy, our credential is not. Reads (blocks_info, account_info, process) all
# work on nano.to's keyless tier, so the client drops the bad key and re-tries
# the SAME endpoint keyless instead of failing the call.
_AUTH_ERRORS = (
    "invalid api key",
    "api key",
    "unauthorized",
    "invalid key",
)


def _is_auth_error(data) -> bool:
    if not isinstance(data, dict):
        return False
    return any(k in str(data.get("error", "")).lower() for k in _AUTH_ERRORS)


def _load_endpoints() -> List[str]:
    """Build endpoint list from NANO_RPC_URL env var.

    Only https://rpc.nano.to is used. Fallback endpoints are intentionally
    absent so the RPC key never reaches a third-party node and an outage is
    never hidden behind a silent reroute.
    """
    raw = os.environ.get("NANO_RPC_URL", "").strip()
    if raw:
        endpoints = [u.strip().rstrip("/") for u in raw.split(",") if u.strip()]
    else:
        endpoints = list(DEFAULT_RPC_URLS)
    return endpoints


def _load_key() -> Optional[str]:
    """Read NANO_RPC_KEY, falling back to NANO_RPC_API_KEY."""
    for name in ("NANO_RPC_KEY", "NANO_RPC_API_KEY"):
        key = os.environ.get(name, "").strip()
        if key:
            return key
    return None


class NanoRPC:
    def __init__(
        self,
        endpoints: Optional[List[str]] = None,
        api_key: Optional[str] = None,
        timeout: float = 20.0,
        retries: int = 3,
    ):
        self.endpoints = list(endpoints) if endpoints else _load_endpoints()
        self.api_key = api_key if api_key is not None else _load_key()
        self.timeout = timeout
        self.retries = retries
        self._last_good_index = 0
        # Flipped to True the first time rpc.nano.to rejects the key, so every
        # later call goes out keyless instead of burning a round-trip on it.
        self.key_rejected = False
        self.last_error: Optional[str] = None

        self.session = requests.Session()
        self.session.headers["User-Agent"] = "vela-v2/0.1"
        adapter = HTTPAdapter(pool_connections=8, pool_maxsize=32, max_retries=0)
        self.session.mount("http://", adapter)
        self.session.mount("https://", adapter)

    def _transport_error(self, data) -> bool:
        """Return True if the response indicates an endpoint/transport problem."""
        if not isinstance(data, dict):
            return True
        err = str(data.get("error", "")).lower()
        if not err:
            return False
        return not any(k in err for k in _NANO_ERRORS)

    def call(self, action: str, params: dict) -> dict:
        """Call a Nano RPC action, trying endpoints and retries."""
        last_err = None

        for attempt in range(1, self.retries + 1):
            # Start from the last known-good endpoint.
            order = [self._last_good_index] + [
                i for i in range(len(self.endpoints)) if i != self._last_good_index
            ]
            for idx in order:
                endpoint = self.endpoints[idx]
                # The API key belongs to rpc.nano.to ONLY — never sent to the
                # fallback (rpc.nano.to accepts it in body or header). Once the
                # key has been rejected we stop attaching it entirely.
                use_key = (
                    bool(self.api_key)
                    and endpoint == PRIMARY_RPC_URL
                    and not self.key_rejected
                )
                try:
                    data = self._post(endpoint, action, params, use_key)
                    if use_key and _is_auth_error(data):
                        # The endpoint is alive and told us the credential is
                        # bad. Retry it immediately keyless rather than falling
                        # through to a different node.
                        self.key_rejected = True
                        print(
                            f"nano_rpc: {endpoint} rejected NANO_RPC_KEY "
                            f"({data.get('error')!r}); continuing keyless. "
                            "Renew the key to restore the paid tier."
                        )
                        data = self._post(endpoint, action, params, use_key=False)
                    if self._transport_error(data):
                        last_err = RuntimeError(f"{endpoint} unusable: {data}")
                        continue
                    self._last_good_index = idx
                    return data
                except Exception as e:
                    # Keep the primary's failure distinguishable from the
                    # fallback's: a masked primary error is what made the
                    # "invalid deposit/commit pair" stall impossible to read.
                    last_err = RuntimeError(f"{endpoint}: {e}")
                    continue

        self.last_error = str(last_err)
        raise RuntimeError(f"All Nano RPC endpoints failed: {last_err}")

    def _post(self, endpoint: str, action: str, params: dict, use_key: bool) -> dict:
        payload = {"action": action, **params}
        headers = {"Content-Type": "application/json"}
        if use_key:
            payload["key"] = self.api_key
            headers["Authorization"] = self.api_key
        resp = self.session.post(
            endpoint, json=payload, headers=headers, timeout=self.timeout
        )
        resp.raise_for_status()
        return resp.json()
